src/beamlime/applications/groups.py

```diff
@@ -0,0 +1,225 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+import asyncio
+from logging import Logger
+from typing import Callable, Optional, TypeVar
+
+from .. import protocols
+from ..core.schedulers import async_timeout
+from .interfaces import BeamlimeApplicationInterface
+from .mixins import LogMixin
+
+_CommunicationChannel = TypeVar(
+    "_CommunicationChannel",
+    protocols.BeamlimeDownstreamProtocol,
+    protocols.BeamlimeUpstreamProtocol,
+    protocols.BeamlimeTwoWayProtocol,
+)
+
+MAX_INSTANCE_NUMBER = 2
+
+
+class ApplicationStoppedException(Exception):
+    ...
+
+
+class ApplicationNotStartedException(Exception):
+    ...
+
+
+# TODO: Implement multi-process/multi-machine instance group interface.
+class BeamlimeApplicationInstanceGroup(LogMixin):
+    """
+    Multiple instances of a single type of application.
+    It overwrites all interfaces of ``BeamlimeApplicationInterface``,
+    in a way that it has the exact same interfaces
+    just like a single application but controls multiple instances internally.
+    Therefore it does not inherit ``BeamlimeApplicationInterface``,
+    but it should pass the protocol test.
+    See ``protocol_test.test_instace_group_protocol`` of the unit test package.
+
+    Parameters
+    ----------
+    constructor: Callable
+        ``constructor`` should be able to construct an instance
+        by calling it without any arguments.
+        It can be made by ``functools.partial``.
+        See the example usage in the ``BeamlimeSystem.build_instances``.
+    instance_num: int
+        Number of instance the ``ApplicationInstanceGroup`` will populate and carry.
+        It initially populates instances of ``instance_num``.
+    app_name: str
+        Name of the application.
+
+    Protocol
+    --------
+    BeamlimeApplicationProtocol
+
+    """
+
+    def __init__(
+        self,
+        constructor: Callable,
+        instance_num: int = 2,
```

Comment:
Is 2 a good default? From what I can see below, you currently cannot add and remove instances dynamically (like appending and popping from a list or a dict of instances).
Do you want the `__init__` to actually create the instances, or do you want the group to just take in as input instances that are already created?

My experience with groups structures is that they end up being more flexible if they just take in things that are already created. You can still perform some initialization inside the group `__init__`, but sometimes you may want to take in some instances that you want to initialize slightly differently, and then if you are creating them inside the group `__init__` that becomes more difficult to achieve.

Reply:
Yes...! I updated the default value to be 1. I forgot to update them after I tested this code : D....

And also I agreed that we shouldn't create the instances at the `__init__` so I removed that part and let the `BeamlimeSystem` do the job. It has the method `init_instances`.
Also, the `populate` method can create more instances.